### PR TITLE
Add compute-vsd tool

### PR DESCRIPTION
**Pull request by jafyvilla**
Modified vsdLoader based on user request.
Move dye curve parameter from URIHandler to compute-vsd tool.

### CODE REVIEW

**Comment by jafyvilla** [comment]
A user guide will follow, as well as a validation test.
This needs to be deployed as a module as soon as possible.

File: apps/computeVSD/computeVSD.cpp [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+public:
+    ComputeVSD()
+        : CommandLineApplication( "Compute the VSD signal and voxelize it to "
+                                  "generate a 2D image through a Beer-Lambert "
+                                  "projection (VTK format)" )
+        , _outputFile( "image" )
+    {
+        _options.add_options()
+//! [ComputeVSDParameters] @anchor ComputeVSD
+            ( "datatype,d", po::value< std::string >()->default_value("float"),
```
Comment (tribal-tec) [inline apps/computeVSD/computeVSD.cpp]: rm if not used

File: apps/computeVSD/computeVSD.cpp [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+            ( "depth", po::value< float >()->default_value( 2081.75641787f ),
+              "Depth of the attenuation curve area of influence. It also "
+              "defines the Y-coordinate at which it starts being applied, "
+              "down until y=0 (default: 2081.75641787 micrometers)." )
+            ( "interpolate-attenuation",
+              "Interpolate the attenuation from the dye curve." )
+            ( "sigma", po::value< double >()->default_value( 0.0045 ),
+              "Absorption + scattering coefficient (units per micrometer) in "
+              "the Beer-Lambert law. Must be a positive value." )
+            ( "v0", po::value< float >()->default_value( 0.f ),
```
Comment (tribal-tec) [inline apps/computeVSD/computeVSD.cpp]: -65 as a more sensible default?

File: fivox/attenuationCurve.h [hunk 3]
```diff
@@ -65,31 +66,39 @@ class AttenuationCurve
...
-        const float diffToIndRatio = (invertedDepth - deltaDepthPerAttn * ind)
-                                                        / deltaDepthPerAttn;
-        return diffToIndRatio * _dyeCurve[ ind + 1 ] +
-               ( 1.0f - diffToIndRatio ) * _dyeCurve[ ind ];
+        const size_t ind = depth / deltaDepthPerAttn;
+        const float diffToIndRatio = (depth - deltaDepthPerAttn * ind)
+                                     / deltaDepthPerAttn;
+
+        const float attenuation = _dyeCurve[ ind + 1 ];
+        return interp ? diffToIndRatio * attenuation +
```
Comment (tribal-tec) [inline fivox/attenuationCurve.h]: ind, diffToIndRation and interp cannot be real words for clarity?

File: tests/uriHandler.cpp [hunk 4]
```diff
@@ -121,6 +121,5 @@ BOOST_AUTO_TEST_CASE(vsd)
     BOOST_CHECK_EQUAL( handler.getType(), fivox::VolumeType::vsd );
     BOOST_CHECK_EQUAL( handler.getInputRange(),
                        fivox::Vector2f( -100000.f, 300.f ));
-    BOOST_CHECK_EQUAL( handler.getDyeCurve(), "" );
```
Comment (tribal-tec) [inline tests/uriHandler.cpp]: test getAreas() instead?

**Comment by tribal-tec** [comment]
With this new review feature, having a separate is more desired as the changes can be tracked and correlated to the review comments. Force push should be avoided, especially you can now squash everything while merging.

File: apps/computeVSD/beerLambertProjectionImageFilter.h [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    {
+        m_Sum = itk::NumericTraits< TAccumulate >::ZeroValue();
+        m_Values.clear();
+    }
+
+    inline void operator()( const TInputPixel& input )
+    {
+        // assume that values are being computed in order (top-down)
+        const double depth =
+                m_PixelSize * ( m_Size - m_Values.size( )) / (double)m_Size;
```
Comment (hernando) [inline apps/computeVSD/beerLambertProjectionImageFilter.h]: This computation yields values between 0 and m_PixelSize, but not between 0 and the volume depth in Y.

File: apps/computeVSD/computeVSD.cpp [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+typedef itk::Image< FloatPixelType, 2 > FloatImageType;
+}
+
+class ComputeVSD : public CommandLineApplication
+{
+public:
+    ComputeVSD()
+        : CommandLineApplication( "Compute the VSD signal and voxelize it to "
+                                  "generate a 2D image through a Beer-Lambert "
+                                  "projection (VTK format)" )
```
Comment (hernando) [inline apps/computeVSD/computeVSD.cpp]: This message are usually descriptions of what the tool is or does, I wouldn't recommend to use imperative. I haven't seen it before. What about: "A tool to compute 2D projections of VSD signals using the Beer-Lambert law."

File: apps/computeVSD/computeVSD.cpp [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+        , _outputFile( "image" )
+    {
+        _options.add_options()
+//! [ComputeVSDParameters] @anchor ComputeVSD
+            ( "output,o", po::value< std::string >(),
+              "Name of the output volume file; contains frame number if "
+              "--frames or --times" )
```
Comment (hernando) [inline apps/computeVSD/computeVSD.cpp]: If the output file format is predefined, mention it here (not in the general description).

File: apps/computeVSD/computeVSD.cpp [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+            ( "export-volume",
+              "Export also the 3d volume containing the VSD values, "
+              "in addition to the VTK file." )
+            ( "sensor-res", po::value< size_t >()->default_value( 512 ),
+              "Number of pixels per side of square sensor." )
+            ( "sensor-dim", po::value< size_t >()->default_value( 1000 ),
+              "Length of side of square sensor in micrometers." )
+            ( "curve", po::value< std::string >(),
+              "Path to the dye curve file (default: no file; attenuation of 1)")
```
Comment (hernando) [inline apps/computeVSD/computeVSD.cpp]: no attenuation instead of attenuation of 1?

File: apps/computeVSD/computeVSD.cpp [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+            ( "sigma", po::value< double >()->default_value( 0.0045 ),
+              "Absorption + scattering coefficient (units per micrometer) in "
+              "the Beer-Lambert law. Must be a positive value "
+              "(default: 0.0045).")
+            ( "v0", po::value< float >()->default_value( -65.f ),
+              "Resting potential (default: -65 mV." )
+            ( "g0", po::value< float >()->default_value( 0.f ),
+              "Multiplier for surface area in background fluorescence term." )
+            ( "ap-threshold", po::value< float >(),
+              "Action potential threshold in millivolts." );
```
Comment (hernando) [inline apps/computeVSD/computeVSD.cpp]: This descriptions are a bit terse but I guess that longer decriptions should go into a some other form of documentation.

File: apps/computeVSD/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) BBP/EPFL 2014-2015, Stefan.Eilemann@epfl.ch
+# All rights reserved. Do not distribute without further notice.
```
Comment (eile) [inline apps/computeVSD/CMakeLists.txt]: Shouldn't this be removed, now that it's OSS?

File: fivox/vsdLoader.h [hunk 11]
```diff
@@ -48,10 +48,55 @@ class VSDLoader : public EventSource
...
+     * @param enable true to enable spike filtering; false to use the input
+     * voltages as they are
+     */
+    void setSpikeFilter( bool enable );
+
+    /**
+     * When the spike filtering is enabled, use this to set the action potential
+     * threshold, in millivolts
+     *
+     * @param apThreshold action potential threshold, in millivolts
```
Comment (jafyvilla) [inline fivox/vsdLoader.h]: The spike filtering is an option used to perform experiments, by changing the action potential threshold. It basically clamps the input voltage to the specified value. This kind of things will be detailed in a brief user guide that I will add in a different PR.

File: fivox/attenuationCurve.h [hunk 12]
```diff
@@ -29,6 +29,7 @@
 
 namespace fivox
 {
+
 /** Attentuation curve access for depth-based interpolation. @sa VSDLoader */
 class AttenuationCurve
```
Comment (jafyvilla) [inline fivox/attenuationCurve.h]: I'd rather keep is at Attenuation, at least for now. It is only used for that; the input file containing the values has a custom specific format; and changes/additions attenuation-specific might be done, based on user feedback.

**Comment by jafyvilla** [comment]
All comments addressed, hopefully. Thanks!

**Pull request merged**